donation-tracker/src/utils/dataStore.ts

```diff
@@ -1,316 +1,486 @@
+        this.registerDataUpdate();
+        return;
+      }
+
+      // compare date age (to avoid older data being used as Google load balancers sometimes behave strange)
+      let newDataAvailableOnline = false;
+      if (this.hasLocalData) {
+        this.localData!.requestTime = tmpParsedData.requestTime;
+        // make sure new check time is persisted:
+        localStorage.setItem('donationCache', JSON.stringify(this.localData));
+        this.debugLog('DataStore: Local data present, comparing timestamps');
+        if (this.localData!.timeStamp < tmpParsedData.timeStamp) {
+          this.debugLog('DataStore: Data online is newer');
+          newDataAvailableOnline = true;
+        }
+      } else {
+        this.debugLog('DataStore: No local data available until now');
+        newDataAvailableOnline = true;
+      }
+      // if new data is available and there is content extract the lines
+      if (rows.length > 0 && newDataAvailableOnline) {
+        this.debugLog('DataStore: Data extracting required');
+        tmpParsedData.data = this.extractDataLines(rows, mapping);
+        // save newer data in local storage & class internal store
+        this.debugLog('DataStore: Persisting new data');
+        localStorage.setItem('donationCache', JSON.stringify(tmpParsedData));
+        this.localData = tmpParsedData;
+        this.hasLocalData = true;
+      }
+    }
+    this.isRefreshing = false;
+    // update components
+    this.updateSubscribers.forEach(subscriber => subscriber(this.localData!.requestTime));
+    // register next execution of data update
+    this.registerDataUpdate();
+  }
+
+  /**
+   * Internal function to extract the data from the lines of the CSV file
+   * @param dataRows The array containing the lines of the SV file, including header line
+   * @param mapping The column mapping for the csv file
+   * @returns An array with the DonationItems that were extracted
+   */
+  private extractDataLines(dataRows: string[], mapping: ColumnMapping): Array<DonationItem> {
+    this.debugLog('DataStore: Data extracting data lines from online source');
+    const tmpResult: Array<DonationItem> = [];
+    // we have three simple values (update date does not count) and three localizable ones; so calculate the min number of columns
+    const minColumns = 3 + mapping.languages.length * 3;
+    for (let idx = 2; idx < dataRows.length; idx++) {
+      const tmpLineData = this.parseCsvLine(dataRows[idx].trim());
+      if (tmpLineData.length >= minColumns) {
+        const tmpObject: DonationItem = {
+          article: {},
+          campaignKey: tmpLineData[mapping.campaignId].toString(),
+          neededOverall: Number.parseInt(tmpLineData[mapping.need].toString()),
+          alreadyDonated: Number.parseInt(tmpLineData[mapping.donated].toString()),
+          remainingNeed: Number.parseInt(tmpLineData[mapping.need].toString()) - Number.parseInt(tmpLineData[mapping.donated].toString()),
```

Comment:
```suggestion
          remainingNeed: Math.max(0, Number.parseInt(tmpLineData[mapping.need].toString()) - Number.parseInt(tmpLineData[mapping.donated].toString())),
```
there might be cases of more donations than needed. Maybe the remaining need should then not be negative.